### PR TITLE
Api::DeployGroupsController

### PR DESCRIPTION
**Pull request by apanzerj**
* Adding Api::DeployGroupsController

/cc @zendesk/samson

### Tasks
 - [ ] :+1: from team

### References
 - Jira link:

### Risks
- Level: Low/Med/High

### CODE REVIEW

File: app/controllers/api/deploy_groups_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+class Api::DeployGroupsController < Api::BaseController
+  before_action :ensure_enabled
+
+  def index
+    scope = current_project.present? ? stage.deploy_groups : DeployGroup.all
+    render json: paginate(scope.sort_by(&:natural_order))
```
Comment (grosser) [inline app/controllers/api/deploy_groups_controller.rb]: paginate does not make much sense if we show all ?

File: config/routes.rb [hunk 2]
```diff
@@ -13,10 +13,16 @@
     resources :deploy_groups, only: [:index]
 
     resources :projects, only: [:index] do
+      resources :stages, only: [:index] do
+        member do
```
Comment (grosser) [inline config/routes.rb]: use `resources :deploy_groups, only: [:index]` instead ?

File: app/controllers/api/deploy_groups_controller.rb (hunk 1 above)
Comment (apanzerj) [inline app/controllers/api/deploy_groups_controller.rb]: wat? paginate breaks them into clumps of 1000.

if there is less than 1000, then yes, paginate is pointless, but >1000 and it doesn't show them all.

File: test/controllers/api/deploy_groups_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+               to: "api/deploy_groups#index", params: {project_id: "1", id: "2"}
+
+  oauth_setup!
+
+  it 'returns precondition_failed when deploy_groups are disabled' do
+    get :index
+    assert_response :precondition_failed
+  end
+
+  describe 'when deploy_groups are enabled' do
```
Comment (grosser) [inline test/controllers/api/deploy_groups_controller_test.rb]: can avoid the wrapping describe with:

```Ruby
before { DeployGroup.stubs(:enabled?).returns(true) }

it 'returns precondition_failed when deploy_groups are disabled' do
  DeployGroup.unstub(:enabled?)
  ...
end


describe "#index" ...
```

File: config/routes.rb (hunk 2 above)
Comment (apanzerj) [inline config/routes.rb]: Yeah, kind of worried about that, I don't want to do three levels of resources foo.

**Comment by grosser** [comment]
👍

File: app/controllers/api/deploy_groups_controller.rb (hunk 1 above)
Comment (grosser) [inline app/controllers/api/deploy_groups_controller.rb]: we still fetch them from the db ... but I guess we don't have to serialize them ... I guess nobody has >1000 ... so fine either way ...

File: test/controllers/api/deploy_groups_controller_test.rb (hunk 3 above)
Comment (apanzerj) [inline test/controllers/api/deploy_groups_controller_test.rb]: unstub. neat.

File: app/controllers/api/deploy_groups_controller.rb (hunk 1 above)
Comment (apanzerj) [inline app/controllers/api/deploy_groups_controller.rb]: I see. Hmm. Yeah, maybe skip the paginate anyways? I think we only need it for things like deploys which are definitely more than 1000?

Comment (grosser) [inline app/controllers/api/deploy_groups_controller.rb]: nice for consistency I guess ...

**Pull request merged**